File: backend/app/models/post.py
from datetime import datetime, timezone
import uuid
from sqlalchemy import String, Text, DateTime, ForeignKey, Enum as SAEnum
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.dialects.postgresql import UUID

from app.core.database import Base
from typing import TYPE_CHECKING

from app.models.enums import PostStatus

if TYPE_CHECKING:
    from app.models.bookmark import Bookmark
    from app.models.category import Category
    from app.models.comment import Comment
    from app.models.like import Like
    from app.models.post_image import PostImage
    from app.models.tag import Tag
    from app.models.user import User

class Post(Base):
    __tablename__ = "posts"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)

    title: Mapped[str] = mapped_column(String, nullable=False)
    slug: Mapped[str] = mapped_column(String, unique=True, nullable=False)

    excerpt: Mapped[str | None] = mapped_column(Text)
    content: Mapped[str] = mapped_column(Text, nullable=False)

    cover_image: Mapped[str | None] = mapped_column(Text)
    cover_image_public_id: Mapped[str | None] = mapped_column(Text)

    status: Mapped[PostStatus] = mapped_column(SAEnum(PostStatus), nullable=False)

    author_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False
    )

    category_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("categories.id", ondelete="RESTRICT"),
        nullable=False
    )

    published_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    created_at: Mapped[datetime] = mapped_column(DateTime, default=lambda: datetime.now(timezone.utc), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
        nullable=False,
    )

    # Relationships
    author: Mapped["User"] = relationship(back_populates="posts")
    category: Mapped["Category"] = relationship(back_populates="posts")

    comments: Mapped[list["Comment"]] = relationship(back_populates="post", cascade="all, delete")
    likes: Mapped[list["Like"]] = relationship(back_populates="post", cascade="all, delete")
    bookmarks: Mapped[list["Bookmark"]] = relationship(back_populates="post", cascade="all, delete")
    tags: Mapped[list["Tag"]] = relationship(secondary="post_tags", back_populates="posts")
    images: Mapped[list["PostImage"]] = relationship(
        back_populates="post",
        cascade="all, delete-orphan",
        order_by="PostImage.position",
    )

File: web/src/types/post.ts
export type PostStatus = "Draft" | "Published" | "Archived"

export type PostAuthor = {
  id: string
  first_name: string
  last_name: string
  username: string
  avatar?: string | null
  name?: string
}

export type PostCategory = {
  id: string
  name: string
  slug: string
}

export type PostTag = {
  id: string
  name: string
  slug: string
}

export type PostImage = {
  id: string
  image_url: string
  alt_text: string
  position?: number | null
  created_at: string
}

export type Post = {
  id: string
  title: string
  slug: string
  excerpt?: string | null
  content: string
  cover_image?: string | null
  status: PostStatus
  author_id: string
  category_id: string
  author: PostAuthor
  category: PostCategory
  tags: PostTag[]
  images: PostImage[]
  published_at?: string | null
  created_at: string
  updated_at: string
}

export type Category = PostCategory & {
  created_at?: string
  updated_at?: string
}

export type Tag = PostTag & {
  created_at?: string
  updated_at?: string
}
